src/vsr/replica.zig

```diff
@@ -440,77 +440,69 @@ pub fn ReplicaType(
-                    assert(!self.journal.faulty.bit(slot));
-                } else if (self.journal.header_for_op(header.op)) |journal_header| {
-                    assert(!self.journal.dirty.bit(slot));
-                    assert(!self.journal.faulty.bit(slot));
+            var op_head: ?u64 = null;
 
-                    if (header.op < journal_header.op) {
-                        // Don't overwrite a newer op.
-                        // (This must be a SV message because a DVC would not have a newer op).
-                        assert(self.log_view == self.view);
-                    } else {
-                        self.journal.set_header_as_dirty(header);
+            if (self.log_view == self.view) {
+                for (self.journal.headers) |*header| {
+                    if (header.command == .prepare) {
+                        assert(header.op <= self.op_checkpoint_trigger());
+                        assert(header.view <= self.log_view);
+
+                        if (op_head == null or op_head.? < header.op) op_head = header.op;
                     }
-                } else {
-                    assert(self.journal.dirty.bit(slot) == self.journal.faulty.bit(slot));
-
-                    self.journal.headers[slot.index] = header.*;
-                    self.journal.dirty.set(slot);
-                    // Don't touch faulty — if it is set, we don't want to unset it. The WAL slot
-                    // may contain a corrupt version is this op, and we don't want to incorrectly
-                    // nack it. (This is why we do not call replace_header()/set_header_as_dirty()
-                    // here.)
                 }
-            }
-
-            // If we were a lagging backup that installed an SV but didn't finish fast-forwarding,
-            // the vsr_headers head op may be part of the checkpoint after this one.
-            maybe(vsr_headers.slice[0].op > self.op_checkpoint_trigger());
+            } else {
+                // Fall-through to choose op-head from vsr_headers.
+                //
+                // "Highest op from log_view in WAL" is not the correct choice for op-head when
+                // recovering with a durable DVC (though we still resort to this if there are no
+                // usable headers in the vsr_headers). It is possible that we started the view and
+                // finished some repair before updating our view_durable.
+                //
+                // To avoid special-casing this all over, we pretend this higher op doesn't
+                // exist. This is safe because we never prepared any ops in the view we joined just
+                // before the crash.
+                assert(self.log_view < self.view);
+                maybe(self.journal.op_maximum() > vsr_headers.slice[0].op);
+            }
+
+            // Try to use vsr_headers to update our head op.
+            // Loading headers prior to the head would risk the following scenario:
```

Comment:
This comment is a bit confusing (at least, I didn't get it until I also read the commit message). Maybe 

```
            // Try to use vsr_headers to update our head op and its header.
            // Don't load headers prior to the head to avoid the following scenario:
```